### PR TITLE
spring remote debug

### PR DESCRIPTION
**Pull request by ivywei0125**
What does this implement/fix? Explain your changes.
---------------------------------------------------
…

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------
…

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java [hunk 1]
```diff
@@ -149,4 +151,23 @@ public SpringCloudPersistentDisk getPersistentDisk() {
...
+        DeploymentsClient client = Utils.getDeploymentsClient(this.getSubscriptionId());
+        client.disableRemoteDebugging(this.getResourceGroupName(), Objects.requireNonNull(this.getParent().getRemote()).name(),
+                this.getName(), this.getActiveDeploymentName());
+    }
+
+    public boolean isRemoteDebuggingEnabled() {
+        DeploymentsClient client = Utils.getDeploymentsClient(this.getSubscriptionId());
+        return client.getRemoteDebuggingConfig(this.getResourceGroupName(), Objects.requireNonNull(this.getParent().getRemote()).name(),
+                this.getName(), this.getActiveDeploymentName()).enabled();
+    }
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudApp.java]: these methods should be moved to `SpringAppDeployment`, 
It already have a method to get the client. there is no need to make a new `Utils.getDeploymentClient`.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudAppInstance.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    public String getDiscoveryStatus() {
+        final DeploymentInstance deploymentInstance = this.getRemote();
+        if (Objects.nonNull(deploymentInstance)) {
+            return deploymentInstance.discoveryStatus();
+        }
+        return Status.UNKNOWN;
+    }
+
+    public String getRemoteUrl() {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudAppInstance.java]: rename as `getRemoteDebuggingUrl` to clearly describe what it is?

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/Utils.java [hunk 3]
```diff
@@ -78,4 +84,14 @@ public static <T> T pollUntil(Callable<T> callable, @Nonnull Predicate<T> predic
...
+
+    public static DeploymentsClient getDeploymentsClient(String subscriptionId) {
+        final Account account = Azure.az(AzureAccount.class).account();
+        final AzureProfile azureProfile = new AzureProfile(account.getEnvironment());
+        return (new AppPlatformManagementClientBuilder())
+                .pipeline(HttpPipelineProvider.buildHttpPipeline(account.getTokenCredential(subscriptionId), azureProfile))
+                .endpoint(azureProfile.getEnvironment().getResourceManagerEndpoint())
+                .subscriptionId(subscriptionId)
+                .buildClient().getDeployments();
+    }
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/Utils.java]: remove this method.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java [hunk 4]
```diff
@@ -34,22 +33,28 @@
...
      */
     protected SpringCloudDeployment(@Nonnull SpringCloudDeployment origin) {
         super(origin);
+        this.instanceModule = origin.instanceModule;
     }
 
     protected SpringCloudDeployment(@Nonnull SpringAppDeployment remote, @Nonnull SpringCloudDeploymentModule module) {
         super(remote.name(), module);
+        this.instanceModule = new SpringCloudAppInstanceModule(this);
     }
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java:61]: you should also override method `getSubModules()`

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java [hunk 5]
```diff
@@ -174,9 +179,11 @@ public Map<String, String> getEnvironmentVariables() {
...
 
-    @Nonnull
-    public List<DeploymentInstance> getInstances() {
-        return Optional.ofNullable(this.getRemote()).map(SpringAppDeployment::instances).orElse(Collections.emptyList());
+    public List<SpringCloudAppInstance> getInstances() {
+        return Optional.ofNullable(this.getRemote()).map(SpringAppDeployment::instances)
+                .orElse(Collections.emptyList()).stream()
+                .map(deploymentInstance -> new SpringCloudAppInstance(deploymentInstance.name(), this.instanceModule))
+                .collect(Collectors.toList());
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudDeployment.java]: why not ?
```
return this.instancesModule.list();
```

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudCluster.java [hunk 6]
```diff
@@ -80,4 +82,8 @@ public String getSku() {
     public boolean isEnterpriseTier() {
         return this.remoteOptional().map(SpringService::sku).filter(s -> s.name().equalsIgnoreCase(SkuName.E0.toString())).isPresent();
     }
+
+    public String getFqdn() {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/SpringCloudCluster.java:87]: `@Nonnull`

**Review by Flanker32 — APPROVED** [review]
LGTM

**Pull request merged**